File: sdk/python/polar/customer_portal/customers.py
from __future__ import annotations

import typing

from polar.base import (
    AsyncServiceBase,
    SyncServiceBase,
    parse_response_json,
    parse_response_none,
)
from polar.errors import (
    CheckEmailUpdate401Error,
    CustomerNotReady,
    HTTPValidationError,
    PaymentMethodInUseByActiveSubscription,
    PaymentMethodSetupFailed,
    ResourceNotFound,
    VerifyEmailUpdate401Error,
    VerifyEmailUpdate422Error,
)
from polar.inputs import (
    CustomerEmailUpdateRequest,
    CustomerEmailUpdateVerifyRequest,
    CustomerPaymentMethodConfirm,
    CustomerPaymentMethodCreate,
    CustomerPortalCustomerUpdate,
)
from polar.outputs import (
    CustomerEmailUpdateVerifyResponse,
    CustomerPaymentMethod,
    CustomerPaymentMethodCreateResponse,
    CustomerPortalCustomer,
    ListResourceCustomerPaymentMethod,
)


class CustomersSync(SyncServiceBase):
    def get(
        self,
    ) -> CustomerPortalCustomer:
        """
        Get authenticated customer.

        **Scopes**: `customer_portal:read` `customer_portal:write`

        Args:

        Raises:
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/customer-portal/customers/me",
            path_params={},
            query_params={},
        )
        response = self.client.send_request(request)
        return parse_response_json(response, CustomerPortalCustomer)

    def update(
        self,
        **kwargs: typing.Unpack[CustomerPortalCustomerUpdate],
    ) -> CustomerPortalCustomer:
        """
        Update authenticated customer.

        Args:
            **kwargs: Request body parameters

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/customer-portal/customers/me",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, CustomerPortalCustomer, method_errors)

    def list_payment_methods(
        self,
        *,
        page: int = 1,
        limit: int = 10,
    ) -> ListResourceCustomerPaymentMethod:
        """
        Get saved payment methods of the authenticated customer.

        Args:
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/customer-portal/customers/me/payment-methods",
            path_params={},
            query_params={
                "page": page,
                "limit": limit,
            },
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(
            response, ListResourceCustomerPaymentMethod, method_errors
        )

    def iter_list_payment_methods(
        self,
        *,
        page: int = 1,
        limit: int = 10,
    ) -> typing.Generator[CustomerPaymentMethod]:
        """
        Get saved payment methods of the authenticated customer.

        Args:
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.

        Returns:
            A generator that yields items of type CustomerPaymentMethod.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        while True:
            response = self.list_payment_methods(
                page=page,
                limit=limit,
            )
            yield from response.items
            if page == response.pagination.max_page:
                break
            page += 1

    def add_payment_method(
        self,
        **kwargs: typing.Unpack[CustomerPaymentMethodCreate],
    ) -> CustomerPaymentMethodCreateResponse:
        """
        Add a payment method to the authenticated customer.

        Args:
            **kwargs: Request body parameters

        Raises:
            PaymentMethodSetupFailed: The card was declined while setting up the payment method.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/customer-portal/customers/me/payment-methods",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            400: PaymentMethodSetupFailed,
            422: HTTPValidationError,
        }
        return parse_response_json(
            response, CustomerPaymentMethodCreateResponse, method_errors
        )

    def confirm_payment_method(
        self,
        **kwargs: typing.Unpack[CustomerPaymentMethodConfirm],
    ) -> CustomerPaymentMethodCreateResponse:
        """
        Confirm a payment method for the authenticated customer.

        Args:
            **kwargs: Request body parameters

        Raises:
            CustomerNotReady: Customer is not ready to confirm a payment method.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/customer-portal/customers/me/payment-methods/confirm",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            400: CustomerNotReady,
            422: HTTPValidationError,
        }
        return parse_response_json(
            response, CustomerPaymentMethodCreateResponse, method_errors
        )

    def delete_payment_method(
        self,
        id: str,
    ) -> None:
        """
        Delete a payment method from the authenticated customer.

        Args:
            id:

        Raises:
            PaymentMethodInUseByActiveSubscription: Payment method is used by active subscription(s).
            ResourceNotFound: Payment method not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="DELETE",
            url="/v1/customer-portal/customers/me/payment-methods/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = self.client.send_request(request)
        method_errors = {
            400: PaymentMethodInUseByActiveSubscription,
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response_none(response, method_errors)

    def request_email_update(
        self,
        **kwargs: typing.Unpack[CustomerEmailUpdateRequest],
    ) -> typing.Any:
        """
        Request an email change for the authenticated customer.

        Args:
            **kwargs: Request body parameters

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/customer-portal/customers/me/email-update/request",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, typing.Any, method_errors)

    def check_email_update(
        self,
        *,
        token: str,
    ) -> None:
        """
        Check if an email change verification token is still valid.

        Args:
            token:

        Raises:
            CheckEmailUpdate401Error: Invalid or expired verification token.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/customer-portal/customers/me/email-update/check",
            path_params={},
            query_params={
                "token": token,
            },
        )
        response = self.client.send_request(request)
        method_errors = {
            401: CheckEmailUpdate401Error,
            422: HTTPValidationError,
        }
        return parse_response_none(response, method_errors)

    def verify_email_update(
        self,
        **kwargs: typing.Unpack[CustomerEmailUpdateVerifyRequest],
    ) -> CustomerEmailUpdateVerifyResponse:
        """
        Verify an email change using the token from the verification email.

        Args:
            **kwargs: Request body parameters

        Raises:
            VerifyEmailUpdate401Error: Invalid or expired verification token.
            VerifyEmailUpdate422Error: Email address is already in use.
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/customer-portal/customers/me/email-update/verify",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            401: VerifyEmailUpdate401Error,
            422: VerifyEmailUpdate422Error,
        }
        return parse_response_json(
            response, CustomerEmailUpdateVerifyResponse, method_errors
        )


class CustomersAsync(AsyncServiceBase):
    async def get(
        self,
    ) -> CustomerPortalCustomer:
        """
        Get authenticated customer.

        **Scopes**: `customer_portal:read` `customer_portal:write`

        Args:

        Raises:
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/customer-portal/customers/me",
            path_params={},
            query_params={},
        )
        response = await self.client.send_request(request)
        return parse_response_json(response, CustomerPortalCustomer)

    async def update(
        self,
        **kwargs: typing.Unpack[CustomerPortalCustomerUpdate],
    ) -> CustomerPortalCustomer:
        """
        Update authenticated customer.

        Args:
            **kwargs: Request body parameters

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/customer-portal/customers/me",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, CustomerPortalCustomer, method_errors)

    async def list_payment_methods(
        self,
        *,
        page: int = 1,
        limit: int = 10,
    ) -> ListResourceCustomerPaymentMethod:
        """
        Get saved payment methods of the authenticated customer.

        Args:
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/customer-portal/customers/me/payment-methods",
            path_params={},
            query_params={
                "page": page,
                "limit": limit,
            },
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(
            response, ListResourceCustomerPaymentMethod, method_errors
        )

    async def iter_list_payment_methods(
        self,
        *,
        page: int = 1,
        limit: int = 10,
    ) -> typing.AsyncGenerator[CustomerPaymentMethod]:
        """
        Get saved payment methods of the authenticated customer.

        Args:
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.

        Returns:
            An async generator that yields items of type CustomerPaymentMethod.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        while True:
            response = await self.list_payment_methods(
                page=page,
                limit=limit,
            )
            for item in response.items:
                yield item
            if page == response.pagination.max_page:
                break
            page += 1

    async def add_payment_method(
        self,
        **kwargs: typing.Unpack[CustomerPaymentMethodCreate],
    ) -> CustomerPaymentMethodCreateResponse:
        """
        Add a payment method to the authenticated customer.

        Args:
            **kwargs: Request body parameters

        Raises:
            PaymentMethodSetupFailed: The card was declined while setting up the payment method.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/customer-portal/customers/me/payment-methods",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            400: PaymentMethodSetupFailed,
            422: HTTPValidationError,
        }
        return parse_response_json(
            response, CustomerPaymentMethodCreateResponse, method_errors
        )

    async def confirm_payment_method(
        self,
        **kwargs: typing.Unpack[CustomerPaymentMethodConfirm],
    ) -> CustomerPaymentMethodCreateResponse:
        """
        Confirm a payment method for the authenticated customer.

        Args:
            **kwargs: Request body parameters

        Raises:
            CustomerNotReady: Customer is not ready to confirm a payment method.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/customer-portal/customers/me/payment-methods/confirm",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            400: CustomerNotReady,
            422: HTTPValidationError,
        }
        return parse_response_json(
            response, CustomerPaymentMethodCreateResponse, method_errors
        )

    async def delete_payment_method(
        self,
        id: str,
    ) -> None:
        """
        Delete a payment method from the authenticated customer.

        Args:
            id:

        Raises:
            PaymentMethodInUseByActiveSubscription: Payment method is used by active subscription(s).
            ResourceNotFound: Payment method not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="DELETE",
            url="/v1/customer-portal/customers/me/payment-methods/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = await self.client.send_request(request)
        method_errors = {
            400: PaymentMethodInUseByActiveSubscription,
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response_none(response, method_errors)

    async def request_email_update(
        self,
        **kwargs: typing.Unpack[CustomerEmailUpdateRequest],
    ) -> typing.Any:
        """
        Request an email change for the authenticated customer.

        Args:
            **kwargs: Request body parameters

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/customer-portal/customers/me/email-update/request",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response_json(response, typing.Any, method_errors)

    async def check_email_update(
        self,
        *,
        token: str,
    ) -> None:
        """
        Check if an email change verification token is still valid.

        Args:
            token:

        Raises:
            CheckEmailUpdate401Error: Invalid or expired verification token.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/customer-portal/customers/me/email-update/check",
            path_params={},
            query_params={
                "token": token,
            },
        )
        response = await self.client.send_request(request)
        method_errors = {
            401: CheckEmailUpdate401Error,
            422: HTTPValidationError,
        }
        return parse_response_none(response, method_errors)

    async def verify_email_update(
        self,
        **kwargs: typing.Unpack[CustomerEmailUpdateVerifyRequest],
    ) -> CustomerEmailUpdateVerifyResponse:
        """
        Verify an email change using the token from the verification email.

        Args:
            **kwargs: Request body parameters

        Raises:
            VerifyEmailUpdate401Error: Invalid or expired verification token.
            VerifyEmailUpdate422Error: Email address is already in use.
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/customer-portal/customers/me/email-update/verify",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            401: VerifyEmailUpdate401Error,
            422: VerifyEmailUpdate422Error,
        }
        return parse_response_json(
            response, CustomerEmailUpdateVerifyResponse, method_errors
        )

File: sdk/typescript/src/2026-04/services/customer_portal/customers.ts
import type { ClientBase } from "../../../base";
import type {
  CustomerEmailUpdateRequest,
  CustomerEmailUpdateVerifyRequest,
  CustomerEmailUpdateVerifyResponse,
  CustomerPaymentMethod,
  CustomerPaymentMethodConfirm,
  CustomerPaymentMethodCreate,
  CustomerPaymentMethodCreateResponse,
  CustomerPortalCustomer,
  CustomerPortalCustomerUpdate,
  ListResourceCustomerPaymentMethod,
} from "../../models";

import {
  CheckEmailUpdate401Error,
  CustomerNotReady,
  HTTPValidationError,
  PaymentMethodInUseByActiveSubscription,
  PaymentMethodSetupFailed,
  ResourceNotFound,
  VerifyEmailUpdate401Error,
  VerifyEmailUpdate422Error,
} from "../../errors";

export const getCustomers = (client: ClientBase) => {
  /**
   * Get authenticated customer.
   *
   * **Scopes**: `customer_portal:read` `customer_portal:write`
   *
   * @returns {CustomerPortalCustomer}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   */
  return async (): Promise<CustomerPortalCustomer> => {
    const pathParams = {};
    const queryParams = {};
    const request = client.buildRequest(
      "GET",
      "/v1/customer-portal/customers/me",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<CustomerPortalCustomer>(response, "json", {});
  };
};
export const updateCustomers = (client: ClientBase) => {
  /**
   * Update authenticated customer.
   *
   * @param body - Request body
   * @returns {CustomerPortalCustomer}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (body: CustomerPortalCustomerUpdate): Promise<CustomerPortalCustomer> => {
    const pathParams = {};
    const queryParams = {};
    const request = client.buildRequest(
      "PATCH",
      "/v1/customer-portal/customers/me",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<CustomerPortalCustomer>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
export const listPaymentMethodsCustomers = (client: ClientBase) => {
  /**
   * Get saved payment methods of the authenticated customer.
   *
   * @param query - Query parameters
   * @returns {ListResourceCustomerPaymentMethod}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (query?: {
    page?: number;
    limit?: number;
  }): Promise<ListResourceCustomerPaymentMethod> => {
    const pathParams = {};
    const queryParams = {
      page: query?.page ?? 1,
      limit: query?.limit ?? 10,
    };
    const request = client.buildRequest(
      "GET",
      "/v1/customer-portal/customers/me/payment-methods",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<ListResourceCustomerPaymentMethod>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
/**
 * Get saved payment methods of the authenticated customer.
 *
 * @param query - Query parameters
 * @returns {AsyncGenerator<CustomerPaymentMethod>} A generator that yields items of type CustomerPaymentMethod.
 * @throws {PolarNetworkError} When a network error occurs
 * @throws {PolarRateLimitError} When the rate limit is exceeded
 * @throws {PolarServerError} When the server returns a 5xx error
 * @throws {HTTPValidationError} Validation Error
 */
export const iterListPaymentMethodsCustomers = (client: ClientBase) => {
  return async function* (query?: {
    page?: number;
    limit?: number;
  }): AsyncGenerator<CustomerPaymentMethod> {
    let page: number;
    page = query?.page ?? 1;
    let limit: number | undefined;
    limit = query?.limit;

    while (true) {
      const response = await listPaymentMethodsCustomers(client)({ ...query, page, limit });
      for (const item of response.items) {
        yield item;
      }
      if (page >= response.pagination.max_page) {
        break;
      }
      page++;
    }
  };
};
export const addPaymentMethodCustomers = (client: ClientBase) => {
  /**
   * Add a payment method to the authenticated customer.
   *
   * @param body - Request body
   * @returns {CustomerPaymentMethodCreateResponse}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {PaymentMethodSetupFailed} The card was declined while setting up the payment method.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    body: CustomerPaymentMethodCreate,
  ): Promise<CustomerPaymentMethodCreateResponse> => {
    const pathParams = {};
    const queryParams = {};
    const request = client.buildRequest(
      "POST",
      "/v1/customer-portal/customers/me/payment-methods",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<CustomerPaymentMethodCreateResponse>(response, "json", {
      400: PaymentMethodSetupFailed,
      422: HTTPValidationError,
    });
  };
};
export const confirmPaymentMethodCustomers = (client: ClientBase) => {
  /**
   * Confirm a payment method for the authenticated customer.
   *
   * @param body - Request body
   * @returns {CustomerPaymentMethodCreateResponse}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {CustomerNotReady} Customer is not ready to confirm a payment method.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    body: CustomerPaymentMethodConfirm,
  ): Promise<CustomerPaymentMethodCreateResponse> => {
    const pathParams = {};
    const queryParams = {};
    const request = client.buildRequest(
      "POST",
      "/v1/customer-portal/customers/me/payment-methods/confirm",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<CustomerPaymentMethodCreateResponse>(response, "json", {
      400: CustomerNotReady,
      422: HTTPValidationError,
    });
  };
};
export const deletePaymentMethodCustomers = (client: ClientBase) => {
  /**
   * Delete a payment method from the authenticated customer.
   *
   * @param id
   * @returns {void}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {PaymentMethodInUseByActiveSubscription} Payment method is used by active subscription(s).
   * @throws {ResourceNotFound} Payment method not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string): Promise<void> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "DELETE",
      "/v1/customer-portal/customers/me/payment-methods/{id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<void>(response, "none", {
      400: PaymentMethodInUseByActiveSubscription,
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};
export const requestEmailUpdateCustomers = (client: ClientBase) => {
  /**
   * Request an email change for the authenticated customer.
   *
   * @param body - Request body
   * @returns {unknown}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (body: CustomerEmailUpdateRequest): Promise<unknown> => {
    const pathParams = {};
    const queryParams = {};
    const request = client.buildRequest(
      "POST",
      "/v1/customer-portal/customers/me/email-update/request",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<unknown>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
export const checkEmailUpdateCustomers = (client: ClientBase) => {
  /**
   * Check if an email change verification token is still valid.
   *
   * @param query - Query parameters
   * @returns {void}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {CheckEmailUpdate401Error} Invalid or expired verification token.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (query: { token: string }): Promise<void> => {
    const pathParams = {};
    const queryParams = {
      token: query.token,
    };
    const request = client.buildRequest(
      "GET",
      "/v1/customer-portal/customers/me/email-update/check",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<void>(response, "none", {
      401: CheckEmailUpdate401Error,
      422: HTTPValidationError,
    });
  };
};
export const verifyEmailUpdateCustomers = (client: ClientBase) => {
  /**
   * Verify an email change using the token from the verification email.
   *
   * @param body - Request body
   * @returns {CustomerEmailUpdateVerifyResponse}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {VerifyEmailUpdate401Error} Invalid or expired verification token.
   * @throws {VerifyEmailUpdate422Error} Email address is already in use.
   */
  return async (
    body: CustomerEmailUpdateVerifyRequest,
  ): Promise<CustomerEmailUpdateVerifyResponse> => {
    const pathParams = {};
    const queryParams = {};
    const request = client.buildRequest(
      "POST",
      "/v1/customer-portal/customers/me/email-update/verify",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<CustomerEmailUpdateVerifyResponse>(response, "json", {
      401: VerifyEmailUpdate401Error,
      422: VerifyEmailUpdate422Error,
    });
  };
};

export function createCustomersService(client: ClientBase) {
  return {
    get: getCustomers(client),
    update: updateCustomers(client),
    listPaymentMethods: listPaymentMethodsCustomers(client),
    addPaymentMethod: addPaymentMethodCustomers(client),
    confirmPaymentMethod: confirmPaymentMethodCustomers(client),
    deletePaymentMethod: deletePaymentMethodCustomers(client),
    requestEmailUpdate: requestEmailUpdateCustomers(client),
    checkEmailUpdate: checkEmailUpdateCustomers(client),
    verifyEmailUpdate: verifyEmailUpdateCustomers(client),
    iterListPaymentMethods: iterListPaymentMethodsCustomers(client),
  };
}

export type Customers = ReturnType<typeof createCustomersService>;
